pkg/proxy/proxy.go

```diff
@@ -0,0 +1,218 @@
+
+	tokenParserInstance, err := auth.DefaultTokenParser()
+	if err != nil {
+		return nil, err
+	}
+	return &Proxy{
+		namespaces:  NewUserNamespaces(app),
+		tokenParser: tokenParserInstance,
+	}, nil
+}
+
+func (p *Proxy) StartProxy() *http.Server {
+	// start server
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", p.handleRequestAndRedirect)
+
+	// listen concurrently to allow for graceful shutdown
+	log.Info(nil, "Starting the Proxy server...")
+	srv := &http.Server{Addr: ":" + ProxyPort, Handler: mux}
+	go func() {
+		if err := srv.ListenAndServe(); err != nil {
+			log.Error(nil, err, err.Error())
+			// TODO: Add a health check and readiness prob for the Proxy
+		}
+	}()
+	return srv
+}
+
+func (p *Proxy) handleRequestAndRedirect(res http.ResponseWriter, req *http.Request) {
+	ctx, err := p.createContext(req)
+	if err != nil {
+		log.Error(nil, err, "unable to create a context")
+		responseWithError(res, crterrors.NewUnauthorizedError("unable to create a context", err.Error()))
+		return
+	}
+	ns, err := p.getTargetNamespace(ctx)
+	if err != nil {
+		log.Error(ctx, err, "unable to get target namespace")
+		responseWithError(res, crterrors.NewInternalError(errors.New("unable to get target namespace"), err.Error()))
+		return
+	}
+
+	p.serveReverseProxy(ctx, ns, res, req)
+}
+
+func responseWithError(res http.ResponseWriter, err *crterrors.Error) {
+	http.Error(res, err.Error(), err.Code)
+}
+
+// createContext creates a new gin.Context with the User ID extracted from the Bearer token.
+// To be used for storing the user ID and logging only.
+func (p *Proxy) createContext(req *http.Request) (*gin.Context, error) {
+	userID, err := p.extractUserID(req)
+	if err != nil {
+		return nil, err
+	}
+	keys := make(map[string]interface{})
+	keys[context.SubKey] = userID
+	return &gin.Context{
+		Keys: keys,
+	}, nil
+}
+
+func (p *Proxy) getTargetNamespace(ctx *gin.Context) (*namespace.Namespace, error) {
+	userID := ctx.GetString(context.SubKey)
+	return p.namespaces.GetNamespace(ctx, userID)
+}
+
+func (p *Proxy) extractUserID(req *http.Request) (string, error) {
+	userToken, err := extractUserToken(req)
+	if err != nil {
+		return "", err
+	}
+
+	token, err := p.tokenParser.FromString(userToken)
+	if err != nil {
+		return "", crterrors.NewUnauthorizedError("unable to extract userID from token", err.Error())
+	}
+	return token.Subject, nil
+}
+
+func extractUserToken(req *http.Request) (string, error) {
+	a := req.Header.Get("Authorization")
+	token := strings.Split(a, "Bearer ")
+	if len(token) < 2 {
+		return "", crterrors.NewUnauthorizedError("no token found", "a Bearer token is expected")
+	}
+	return token[1], nil
+}
+
+func (p *Proxy) serveReverseProxy(ctx *gin.Context, target *namespace.Namespace, res http.ResponseWriter, req *http.Request) {
+	proxy := p.newReverseProxy(ctx, target)
+
+	// Note that ServeHttp is non blocking and uses a go routine under the hood
+	proxy.ServeHTTP(res, req)
+}
```

Comment:
for the sake of readability, I would suggest that these 2 lines were moved in line 93, inside the `handleRequestAndRedirect` func

Reply:
Done in https://github.com/codeready-toolchain/registration-service/pull/224/commits/d5534b9647db7e4c182c10ae074c78e1b524ed83